src/Middleware/HttpLogging/src/W3CLogger.cs

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.Extensions.Options;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Http;
+using System.Numerics;
+using System.Text;
+using Microsoft.Net.Http.Headers;
+using System.Globalization;
+using System.Diagnostics;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.HttpLogging
+{
+    internal class W3CLogger : IAsyncDisposable
+    {
+        private readonly W3CLoggerProcessor _messageQueue;
+        private readonly IOptionsMonitor<W3CLoggerOptions> _options;
+
+        public W3CLogger(IOptionsMonitor<W3CLoggerOptions> options, IHostEnvironment environment, ILoggerFactory factory)
+        {
+            _options = options;
+            _messageQueue = InitializeMessageQueue(_options, environment, factory);
+        }
+
+        // Virtual for testing
+        internal virtual W3CLoggerProcessor InitializeMessageQueue(IOptionsMonitor<W3CLoggerOptions> options, IHostEnvironment environment, ILoggerFactory factory)
+        {
+            return new W3CLoggerProcessor(options, environment, factory);
+        }
+
+        public async ValueTask DisposeAsync() => await _messageQueue.DisposeAsync();
+
+        public void Log(string[] elements)
+        {
+            _messageQueue.EnqueueMessage(Format(elements));
+        }
+
+        private string Format(string[] elements)
```

Comment:
Since we know the only consumer of this data is StreamWriter, can you delay the Format, enqueue the elements array, and format when writing to the stream? This intermediate string is likely large and useless.

Reply:
Good idea - I'll file a follow-up issue for this one